[Type System] Contributor Guide for Type System Concepts (#20120)
strapi/strapi: docs/docs/guides/05-type-system/02-concepts/01-schema.mdx

---
title: Schema
tags:
  - typescript
  - type system
  - type
  - concepts
---

The schema is the primary data structure leveraged within the Strapi Type System, defining how content is structured and managed in the application.

It serves several key functions:

- **Representation**: At its core, a schema outlines and defines the structure of Strapi content. This is useful when dealing with features that need access to low level schema properties (_e.g. attributes, plugin options, etc..._).

- **Inference**: The schema allows inferring and configuring numerous other types. This includes entities like `ContentType` or `Component`, among others.

### Scope

> Schema types represents **loaded** schemas in the context of a Strapi server application and should be used accordingly.

Author: Not the static definition in schema.js or schema.json file, but schema once they've been loaded by the Strapi server (with plugin attributes, etc...)

> Schema types represents **loaded** schemas in the context of a Strapi server application and should be used accordingly.

Author: I'll add one :slightly_smiling_face:

:::caution

> Database models and schema definitions (before being loaded by Strapi) are not the same types and can't be used interchangeably.

Author: I meant that both the database models AND the schema definitions are NOT the same as loaded schema definition, but maybe it's not clear with the sentence :/

:::

### Sub-Types

Each box is a type that extends the base Schema interface.

In between each box is a text that represents the discriminant used to differentiate the subtype from others.

```mermaid
flowchart TB;
  Schema -- "<code>modelType: contentType</code>" ---- ContentTypeSchema
  Schema -- "<code>modelType: component</code>" ----  ComponentSchema
  ContentTypeSchema -- "<code>kind: collectionType</code>" ---- CollectionTypeSchema
  ContentTypeSchema -- "<code>kind: singleType</code>" ---- SingleTypeSchema
```

### Properties

> Schema types contains useful information that helps other types know how to interact with the Strapi content.

Reviewer: ```suggestion
Schema types contain useful information that helps other types know how to interact with the Strapi content.
```

This is facilitated through multiple properties.

#### Options

A set of properties used to configure the schema. It contains information on features activation among other things.

This can be really useful to make the types adapt to a given schema.

For instance, the document service uses the `options.draftAndPublish` property to determine whether it should add publication methods to the service type.

#### Plugin Options

These options provide the ability to alter or enhance the behaviour of the system based on specific values.

If a plugin is enabled, it might bring functionality that can affect how types interact with each other.

For example, it's possible to add or remove certain entity-service filters from the query type based on whether a plugin is enabled.

#### Attributes

> Strongly typed schema attribute allows the Type System to infer actual entities types based on their properties.

Reviewer 2: "A strongly typed schema attribute..." or "Strongly type schema attributes allow..." I think

> Strongly typed schema attribute allows the Type System to infer actual entities types based on their properties.

Reviewer 2: ```suggestion
Strongly type schema attributes allow the Type System to infer actual entities types based on their properties.
```

Author: In this context, we are referring to the schema attributes being strongly typed, but it's missing an 's' on attribute indeed

Author: ```suggestion
Strongly typed schema attributes allows the Type System to infer actual entities types based on their properties.
```

For instance, a string attribute will resolve to a primitive string in an entity, whereas a repeatable component attribute will resolve to an array of objects.

### Usage

import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

<Tabs>
<TabItem value="public" label="Public" default>

> When designing public APIs (and in most other scenario), it's advised to use the high-level schema types found in the `Schema` namespace.

Reviewer 2: "scenarios"

> When designing public APIs (and in most other scenario), it's advised to use the high-level schema types found in the `Schema` namespace.

Reviewer 2: ```suggestion
When designing public APIs (and in most other scenarios), it's advised to use the high-level schema types found in the `Schema` namespace.
```
